In [2]:
# it is not explicit, but I assume we're supposed to be working with the sakila database. 

# Using SQLAlchemy
from sqlalchemy import create_engine

# Database connection settings
host = "localhost"  # host
user = "root"  #  MySQL username
password = "Malcomx1"  #  MySQL password
database = "sakila"  #  database name


# creates a SQLAlchemy engine, which serves as the entry point for interacting with the database
engine = create_engine(f"mysql+mysqlconnector://{user}:{password}@{host}/{database}")

print(f"Connected to the {database} database successfully!")

Connected to the sakila database successfully!


### 1. Calculate the average rental duration (in days) for each film:

In [ ]:
import pandas as pd
from sqlalchemy import create_engine, text

query = text("""
SELECT
    title,
    rental_duration,
    AVG(rental_duration) OVER () AS avg_rental_duration
FROM
    film;
""")

# Use a connection from the engine
with engine.connect() as connection:
    df_question1 = pd.read_sql(query, connection)

# Display the results
print(df_question1)

                 title  rental_duration  avg_rental_duration
0     ACADEMY DINOSAUR                6                4.985
1       ACE GOLDFINGER                3                4.985
2     ADAPTATION HOLES                7                4.985
3     AFFAIR PREJUDICE                5                4.985
4          AFRICAN EGG                6                4.985
..                 ...              ...                  ...
995     YOUNG LANGUAGE                6                4.985
996         YOUTH KICK                4                4.985
997       ZHIVAGO CORE                6                4.985
998  ZOOLANDER FICTION                5                4.985
999          ZORRO ARK                3                4.985

[1000 rows x 3 columns]


|title|rental_duration|avg_rental_duration|
|-----|---------------|-------------------|
|ACADEMY DINOSAUR|6              |4.9850             |
|ACE GOLDFINGER|3              |4.9850             |
|ADAPTATION HOLES|7              |4.9850             |
|AFFAIR PREJUDICE|5              |4.9850             |
|AFRICAN EGG|6              |4.9850             |
|AGENT TRUMAN|3              |4.9850             |
|AIRPLANE SIERRA|6              |4.9850             |
|AIRPORT POLLOCK|6              |4.9850             |
|ALABAMA DEVIL|3              |4.9850             |
|ALADDIN CALENDAR|6              |4.9850             |
|ALAMO VIDEOTAPE|6              |4.9850             |
|ALASKA PHANTOM|6              |4.9850             |
|ALI FOREVER|4              |4.9850             |
|ALICE FANTASIA|6              |4.9850             |


###  2. Calculate the average payment amount for each staff member:

In [7]:
import pandas as pd
from sqlalchemy import create_engine, text

query = text("""
SELECT 
    staff_id, 
    ROUND(AVG(amount), 2) AS avg_payment_amount
FROM 
    payment
GROUP BY 
    staff_id;

""")

# Use a connection from the engine
with engine.connect() as connection:
    df_question2 = pd.read_sql(query, connection)

# Display the results
print(df_question2)


   staff_id  avg_payment_amount
0         1                4.16
1         2                4.25


#### Expected output

|staff_id|avg_payment_amount|
|--------|------------------|
|1       |4.156568          |
|1       |4.156568          |
|1       |4.156568          |
|1       |4.156568          |
|1       |4.156568          |
|1       |4.156568          |
|1       |4.156568          |

### 3. Calculate the total revenue for each customer, showing the running total within each customer's rental history:

In [9]:
import pandas as pd
from sqlalchemy import create_engine, text

query = text("""
SELECT
    payment.customer_id,
    rental_id,
    rental_date,
    amount,
    SUM(amount) OVER (PARTITION BY customer_id ORDER BY rental_date) AS running_total
FROM
    payment
    JOIN rental USING (rental_id)
ORDER BY
    customer_id, rental_date;
""")

# Use a connection from the engine
with engine.connect() as connection:
    df_question3 = pd.read_sql(query, connection)

# Display the results
print(df_question3)


       customer_id  rental_id         rental_date  amount  running_total
0                1         76 2005-05-25 11:30:37    2.99           2.99
1                1        573 2005-05-28 10:35:23    0.99           3.98
2                1       1185 2005-06-15 00:54:12    5.99           9.97
3                1       1422 2005-06-15 18:02:53    0.99          10.96
4                1       1476 2005-06-15 21:08:46    9.99          20.95
...            ...        ...                 ...     ...            ...
16039          599      14599 2005-08-21 17:43:42    4.99          66.85
16040          599      14719 2005-08-21 21:41:57    1.99          68.84
16041          599      15590 2005-08-23 06:09:44    8.99          77.83
16042          599      15719 2005-08-23 11:08:46    2.99          80.82
16043          599      15725 2005-08-23 11:25:00    2.99          83.81

[16044 rows x 5 columns]


#### Expected output

|customer_id|rental_id|rental_date        |amount|running_total|
|-----------|---------|-------------------|------|-------------|
|1          |76       |2005-05-25 11:30:37|2.99  |2.99         |
|1          |573      |2005-05-28 10:35:23|0.99  |3.98         |
|1          |1185     |2005-06-15 00:54:12|5.99  |9.97         |
|1          |1422     |2005-06-15 18:02:53|0.99  |10.96        |
|1          |1476     |2005-06-15 21:08:46|9.99  |20.95        |
|1          |1725     |2005-06-16 15:18:57|4.99  |25.94        |
|1          |2308     |2005-06-18 08:41:48|4.99  |30.93        |
|1          |2363     |2005-06-18 13:33:59|0.99  |31.92        |
|1          |3284     |2005-06-21 06:24:45|3.99  |35.91        |
|1          |4526     |2005-07-08 03:17:05|5.99  |41.90        |

### 4. Determine the quartile for the rental rates of films:

In [10]:

import pandas as pd
from sqlalchemy import create_engine, text

query = text("""
SELECT
    title,
    rental_rate,
    NTILE(4) OVER (ORDER BY rental_rate) AS quartile
FROM
    film;
""")

# Use a connection from the engine
with engine.connect() as connection:
    df_question4 = pd.read_sql(query, connection)

# Display the results
print(df_question4)


                    title  rental_rate  quartile
0        ACADEMY DINOSAUR         0.99         1
1         ALAMO VIDEOTAPE         0.99         1
2          ALASKA PHANTOM         0.99         1
3          ALICE FANTASIA         0.99         1
4              ALONE TRIP         0.99         1
..                    ...          ...       ...
995  WONDERLAND CHRISTMAS         4.99         4
996   WORKING MICROCOSMOS         4.99         4
997         WYOMING STORM         4.99         4
998           YENTL IDAHO         4.99         4
999             ZORRO ARK         4.99         4

[1000 rows x 3 columns]


#### Expected output

|title|rental_rate|quartile           |
|-----|-----------|-------------------|
|ACADEMY DINOSAUR|0.99       |1                  |
|ALAMO VIDEOTAPE|0.99       |1                  |
|ALASKA PHANTOM|0.99       |1                  |
|ALICE FANTASIA|0.99       |1                  |
|ALONE TRIP|0.99       |1                  |
|ALTER VICTORY|0.99       |1                  |
|AMADEUS HOLY|0.99       |1                  |
|ANACONDA CONFESSIONS|0.99       |1                  |
|ANNIE IDENTITY|0.99       |1                  |
|ANONYMOUS HUMAN|0.99       |1                  |


### 5. Determine the first and last rental date for each customer:

In [14]:

import pandas as pd
from sqlalchemy import create_engine, text

query = text("""
SELECT
    customer_id,
    MIN(rental_date) AS first_rental_date,
    MAX(rental_date) AS last_rental_date
FROM
    rental
GROUP BY
    customer_id;
    
""")

# Use a connection from the engine
with engine.connect() as connection:
    df_question5 = pd.read_sql(query, connection)

# Display the results
print(df_question5)




     customer_id   first_rental_date    last_rental_date
0              1 2005-05-25 11:30:37 2005-08-22 20:03:46
1              2 2005-05-27 00:09:24 2005-08-23 17:39:35
2              3 2005-05-27 17:17:09 2005-08-23 07:10:14
3              4 2005-06-15 09:31:28 2005-08-23 07:43:00
4              5 2005-05-29 07:25:16 2006-02-14 15:16:03
..           ...                 ...                 ...
594          595 2005-05-28 15:27:22 2005-08-23 21:26:35
595          596 2005-05-26 21:16:52 2006-02-14 15:16:03
596          597 2005-05-25 04:19:28 2006-02-14 15:16:03
597          598 2005-06-20 10:10:29 2005-08-23 00:44:15
598          599 2005-05-31 01:18:56 2005-08-23 11:25:00

[599 rows x 3 columns]


#### Expected result

|customer_id|first_rental_date|last_rental_date   |
|-----------|-----------------|-------------------|
|1          |2005-05-25 11:30:37|2005-08-22 20:03:46|
|1          |2005-05-25 11:30:37|2005-08-22 20:03:46|
|1          |2005-05-25 11:30:37|2005-08-22 20:03:46|
|1          |2005-05-25 11:30:37|2005-08-22 20:03:46|
|1          |2005-05-25 11:30:37|2005-08-22 20:03:46|
|1          |2005-05-25 11:30:37|2005-08-22 20:03:46|
|1          |2005-05-25 11:30:37|2005-08-22 20:03:46|
|1          |2005-05-25 11:30:37|2005-08-22 20:03:46|
|1          |2005-05-25 11:30:37|2005-08-22 20:03:46|
|1          |2005-05-25 11:30:37|2005-08-22 20:03:46|
|1          |2005-05-25 11:30:37|2005-08-22 20:03:46|
|1          |2005-05-25 11:30:37|2005-08-22 20:03:46|
|1          |2005-05-25 11:30:37|2005-08-22 20:03:46|


### 6. Calculate the rank of customers based on their rental counts:

In [15]:

import pandas as pd
from sqlalchemy import create_engine, text

query = text("""
SELECT
    customer_id,
    rental_count,
    RANK() OVER (ORDER BY rental_count DESC) AS rental_count_rank
FROM (
    SELECT
        customer_id,
        COUNT(rental_id) AS rental_count
    FROM
        rental
    GROUP BY
        customer_id
) AS rental_counts;
""")

# Use a connection from the engine
with engine.connect() as connection:
    df_question6 = pd.read_sql(query, connection)

# Display the results
print(df_question6)




     customer_id  rental_count  rental_count_rank
0            148            46                  1
1            526            45                  2
2            144            42                  3
3            236            42                  3
4             75            41                  5
..           ...           ...                ...
594          248            15                594
595           61            14                596
596          110            14                596
597          281            14                596
598          318            12                599

[599 rows x 3 columns]


#### Expected output

|customer_id|rental_count                 |rental_count_rank|
|-----------|-----------------------------|-----------------|
|148        |46                           |1                |
|526        |45                           |2                |
|144        |42                           |3                |
|236        |42                           |3                |
|75         |41                           |5                |
|197        |40                           |6                |
...|...|...

### 7. Calculate the running total of revenue per day for the 'Family' film category:

In [ ]:
# The expected answer has the film name as the category. 

import pandas as pd
from sqlalchemy import create_engine, text

query = text("""
SELECT
    film.title,
    rental.rental_date,
    payment.amount,
    SUM(payment.amount) OVER (PARTITION BY rental.rental_date ORDER BY rental.rental_date) AS daily_revenue
FROM
    film
JOIN
    inventory ON film.film_id = inventory.film_id
JOIN
    rental ON inventory.inventory_id = rental.inventory_id
JOIN
    payment ON rental.rental_id = payment.rental_id
WHERE
    film.rating = 'G';
""")

# Use a connection from the engine
with engine.connect() as connection:
    df_question7 = pd.read_sql(query, connection)

# Display the results
print(df_question7)




                     title         rental_date  amount  daily_revenue
0          BLANKET BEVERLY 2005-05-24 22:53:30    2.99           2.99
1            GRADUATE LORD 2005-05-24 23:03:39    3.99           3.99
2           LAWLESS VISION 2005-05-24 23:31:46    4.99           4.99
3             HANGING DEEP 2005-05-25 00:02:21    5.99           5.99
4       MONTEREY LABYRINTH 2005-05-25 00:31:15    0.99           0.99
...                    ...                 ...     ...            ...
2768          PULP BEVERLY 2006-02-14 15:16:03    2.99         104.63
2769     SLEEPLESS MONSOON 2006-02-14 15:16:03    4.99         104.63
2770       SUMMER SCARFACE 2006-02-14 15:16:03    0.99         104.63
2771  SWEETHEARTS SUSPECTS 2006-02-14 15:16:03    0.99         104.63
2772          TORQUE BOUND 2006-02-14 15:16:03    5.98         104.63

[2773 rows x 4 columns]


#### Expected ouput

|film_category|rental_date                  |amount|daily_revenue|
|-------------|-----------------------------|------|-------------|
|BLANKET BEVERLY|2005-05-24 22:53:30          |2.99  |2.99         |
|GRADUATE LORD|2005-05-24 23:03:39          |3.99  |3.99         |
|LAWLESS VISION|2005-05-24 23:31:46          |4.99  |4.99         |
|HANGING DEEP |2005-05-25 00:02:21          |5.99  |5.99         |
|MONTEREY LABYRINTH|2005-05-25 00:31:15          |0.99  |0.99         |
|SCISSORHANDS SLUMS|2005-05-25 01:48:41          |2.99  |2.99         |
|EASY GLADIATOR|2005-05-25 03:41:50          |4.99  |4.99         |

### 8. Assign a unique ID to each payment within each customer's payment history:

In [20]:
# The expected answer has the film name as the category. 

import pandas as pd
from sqlalchemy import create_engine, text

query = text("""
SELECT
    customer.customer_id,
    CONCAT('PID_', payment.payment_id) AS payment_id,
    payment.amount,
    payment.payment_date,
    ROW_NUMBER() OVER (PARTITION BY customer.customer_id ORDER BY payment.payment_date) AS payment_sequence_id
FROM
    customer
JOIN
    rental ON customer.customer_id = rental.customer_id
JOIN
    payment ON rental.rental_id = payment.rental_id
ORDER BY
    customer.customer_id, payment.payment_date;
""")

# Use a connection from the engine
with engine.connect() as connection:
    df_question8 = pd.read_sql(query, connection)

# Display the results
print(df_question8)




       customer_id payment_id  amount        payment_date  payment_sequence_id
0                1      PID_1    2.99 2005-05-25 11:30:37                    1
1                1      PID_2    0.99 2005-05-28 10:35:23                    2
2                1      PID_3    5.99 2005-06-15 00:54:12                    3
3                1      PID_4    0.99 2005-06-15 18:02:53                    4
4                1      PID_5    9.99 2005-06-15 21:08:46                    5
...            ...        ...     ...                 ...                  ...
16039          599  PID_16045    4.99 2005-08-21 17:43:42                   15
16040          599  PID_16046    1.99 2005-08-21 21:41:57                   16
16041          599  PID_16047    8.99 2005-08-23 06:09:44                   17
16042          599  PID_16048    2.99 2005-08-23 11:08:46                   18
16043          599  PID_16049    2.99 2005-08-23 11:25:00                   19

[16044 rows x 5 columns]


|customer_id|payment_id|payment_sequence_id|
|-----------|----------|-------------------|
|1          |1         |1                  |
|1          |2         |2                  |
|1          |3         |3                  |
|1          |4         |4                  |
|1          |5         |5                  |
|1          |6         |6                  |
|1          |7         |7                  |
|1          |8         |8                  |
|1          |9         |9                  |
|1          |10        |10                 |
|1          |11        |11                 |
|1          |12        |12                 |
|1          |13        |13                 |
|1          |14        |14                 |
|1          |15        |15                 |
|1          |16        |16                 |
|1          |17        |17                 |
|1          |18        |18                 |
|1          |19        |19                 |
|1          |20        |20                 |
|1          |21        |21                 |
|1          |22        |22                 |
|1          |23        |23                 |
|1          |24        |24                 |
|1          |25        |25                 |
|1          |26        |26                 |
|1          |27        |27                 |
|1          |28        |28                 |
|1          |29        |29                 |
|1          |30        |30                 |
|1          |31        |31                 |
|1          |32        |32                 |
|2          |33        |1                  |
|2          |34        |2                  |
|2          |35        |3                  |
|2          |36        |4                  |
|2          |37        |5                  |
|2          |38        |6                  |
|2          |39        |7                  |
|2          |40        |8                  |
|2          |41        |9                  |
|2          |42        |10                 |
|2          |43        |11                 |
|2          |44        |12                 |


### 9. Calculate the difference in days between each rental and the previous rental for each customer:

In [21]:
# The expected answer has the film name as the category. 

import pandas as pd
from sqlalchemy import create_engine, text

query = text("""
SELECT
    customer.customer_id,
    rental.rental_id,
    rental.rental_date,
    COALESCE(DATEDIFF(rental.rental_date, LAG(rental.rental_date) OVER (
        PARTITION BY customer.customer_id 
        ORDER BY rental.rental_date
    )), 0) AS days_since_last_rental
FROM
    customer
JOIN
    rental ON customer.customer_id = rental.customer_id
ORDER BY
    customer.customer_id, rental.rental_date;

""")

# Use a connection from the engine
with engine.connect() as connection:
    df_question9 = pd.read_sql(query, connection)

# Display the results
print(df_question9)




       customer_id  rental_id         rental_date  days_since_last_rental
0                1         76 2005-05-25 11:30:37                       0
1                1        573 2005-05-28 10:35:23                       3
2                1       1185 2005-06-15 00:54:12                      18
3                1       1422 2005-06-15 18:02:53                       0
4                1       1476 2005-06-15 21:08:46                       0
...            ...        ...                 ...                     ...
16039          599      14599 2005-08-21 17:43:42                       0
16040          599      14719 2005-08-21 21:41:57                       0
16041          599      15590 2005-08-23 06:09:44                       2
16042          599      15719 2005-08-23 11:08:46                       0
16043          599      15725 2005-08-23 11:25:00                       0

[16044 rows x 4 columns]


#### Expected output


|customer_id|rental_id|rental_date        |previous_rental_date|days_between_rentals|
|-----------|---------|-------------------|--------------------|--------------------|
|1          |76       |2005-05-25 11:30:37|NULL                |NULL                |
|1          |573      |2005-05-28 10:35:23|2005-05-25 11:30:37 |3                   |
|1          |1185     |2005-06-15 00:54:12|2005-05-28 10:35:23 |18                  |
|1          |1422     |2005-06-15 18:02:53|2005-06-15 00:54:12 |0                   |
|1          |1476     |2005-06-15 21:08:46|2005-06-15 18:02:53 |0                   |
|1          |1725     |2005-06-16 15:18:57|2005-06-15 21:08:46 |1                   |
|1          |2308     |2005-06-18 08:41:48|2005-06-16 15:18:57 |2                   |
|1          |2363     |2005-06-18 13:33:59|2005-06-18 08:41:48 |0                   |
|1          |3284     |2005-06-21 06:24:45|2005-06-18 13:33:59 |3                   |
|1          |4526     |2005-07-08 03:17:05|2005-06-21 06:24:45 |17                  |
|1          |4611     |2005-07-08 07:33:56|2005-07-08 03:17:05 |0                   |

